src/confseq/betting.py

```diff
@@ -0,0 +1,889 @@
+
+    assert 0 < trunc_scale <= 1
+    # if we want to truncate with m
+    if m_trunc:
+        with np.errstate(divide="ignore"):
+            lambdas_positive = np.minimum(lambdas_positive, trunc_scale / mu_t)
+            lambdas_positive = np.maximum(lambdas_positive, -trunc_scale / (1 - mu_t))
+
+            lambdas_negative = np.minimum(lambdas_negative, trunc_scale / (1 - mu_t))
+            lambdas_negative = np.maximum(lambdas_negative, -trunc_scale / mu_t)
+    else:
+        lambdas_positive = np.minimum(lambdas_positive, trunc_scale)
+        lambdas_positive = np.maximum(lambdas_positive, -trunc_scale)
+
+        lambdas_negative = np.minimum(lambdas_negative, trunc_scale)
+        lambdas_negative = np.maximum(lambdas_negative, -trunc_scale)
+
+    with np.errstate(invalid="ignore"):
+        multiplicand_positive = 1 + lambdas_positive * (x - mu_t)
+        multiplicand_negative = 1 - lambdas_negative * (x - mu_t)
+
+    # Use convention that inf * 0 = 0. We still have
+    # a martingale under the null
+    multiplicand_positive[
+        np.logical_and(lambdas_positive == math.inf, x - mu_t == 0)
+    ] = 1
+    multiplicand_negative[
+        np.logical_and(lambdas_negative == math.inf, x - mu_t == 0)
+    ] = 1
+    
+    with np.errstate(invalid="ignore"):
+        capital_process_positive = np.cumprod(multiplicand_positive)
+        capital_process_negative = np.cumprod(multiplicand_negative)
+
+    # If we get nans from 0 * inf, this should be 0
+    capital_process_positive[np.isnan(capital_process_positive)] = 0
+    capital_process_negative[np.isnan(capital_process_negative)] = 0
+    
+    if theta == 1:
+        capital_process = theta * capital_process_positive
+    elif theta == 0:
+        capital_process = (1 - theta) * capital_process_negative
+    else:
+        if convex_comb:
+            capital_process = (
+                theta * capital_process_positive
+                + (1 - theta) * capital_process_negative
+            )
+        else:
+            capital_process = np.maximum(
+                theta * capital_process_positive, (1 - theta) * capital_process_negative
+            )
+    
+    # If mu_t < 0 or mu_t > 1, we cannot be under the null
+    capital_process[np.logical_or(mu_t < 0, mu_t > 1)] = math.inf
+    
+    assert not any(np.isnan(capital_process))
+
+    assert all(capital_process >= 0)
+
+    return capital_process
+
+
+def betting_cs(
+    x,
+    lambdas_fns_positive=None,
+    lambdas_fns_negative=None,
+    alpha=0.05,
+    N=None,
+    breaks=1000,
+    running_intersection=False,
+    parallel=False,
+    convex_comb=False,
+    theta=1 / 2,
+    trunc_scale=1 / 2,
+    m_trunc=True,
+):
+    """
+    Betting-based confidence sequence
+
+    Parameters
+    ----------
+    x, array-like
+        The vector of observations between 0 and 1.
+
+    alpha, real
+        Significance level between 0 and 1.
+
+    lambdas_fn_postive, bivariate function or None
```

Comment:
these look out of date, should be "fns" and document that it's allowed to be a list.

Reply:
Fixed in #12